avoidance/src/avoidance_node.cpp

```diff
@@ -70,4 +82,82 @@ void AvoidanceNode::checkFailsafe(ros::Duration since_last_cloud, ros::Duration
     }
   }
 }
-}
\ No newline at end of file
+
+void AvoidanceNode::px4ParamsCallback(const mavros_msgs::Param& msg) {
+  // collect all px4_ parameters needed for model based trajectory planning
+  // when adding new parameter to the struct ModelParameters,
+  // add new else if case with correct value type
+  auto parse_param_f = [&msg](const std::string& name, float& val) -> bool {
+    if (msg.param_id == name) {
+      ROS_INFO("parameter %s is set from  %f to %f \n", name.c_str(), val, msg.value.real);
+      val = msg.value.real;
+      return true;
+    }
+    return false;
+  };
+
+  auto parse_param_i = [&msg](const std::string& name, int& val) -> bool {
+    if (msg.param_id == name) {
+      ROS_INFO("parameter %s is set from %i to %li \n", name.c_str(), val, msg.value.integer);
+      val = msg.value.integer;
+      return true;
+    }
+    return false;
+  };
+
+  // clang-format off
+  parse_param_f("MPC_ACC_DOWN_MAX", px4_.param_mpc_acc_down_max) ||
+  parse_param_f("MPC_ACC_HOR", px4_.param_mpc_acc_hor) ||
+  parse_param_f("MPC_ACC_UP_MAX", px4_.param_acc_up_max) ||
+  parse_param_i("MPC_AUTO_MODE", px4_.param_mpc_auto_mode) ||
+  parse_param_f("MPC_JERK_MIN", px4_.param_mpc_jerk_min) ||
+  parse_param_f("MPC_JERK_MAX", px4_.param_mpc_jerk_max) ||
+  parse_param_f("MPC_LAND_SPEED", px4_.param_mpc_land_speed) ||
+  parse_param_f("MPC_TKO_SPEED", px4_.param_mpc_tko_speed) ||
+  parse_param_f("MPC_XY_CRUISE", px4_.param_mpc_xy_cruise) ||
+  parse_param_f("MPC_Z_VEL_MAX_DN", px4_.param_mpc_vel_max_dn) ||
+  parse_param_f("MPC_Z_VEL_MAX_UP", px4_.param_mpc_z_vel_max_up) ||
+  parse_param_f("MPC_COL_PREV_D", px4_.param_mpc_col_prev_d) ||
+  parse_param_f("NAV_ACC_RAD", px4_.param_nav_acc_rad);
+  // clang-format on
+}
+
+void AvoidanceNode::checkPx4Parameters() {
+  auto& client = get_px4_param_client_;
+  auto request_param = [&client](const std::string& name, float& val) {
+    mavros_msgs::ParamGet req;
+    req.request.param_id = name;
+    if (client.call(req) && req.response.success) {
+      val = req.response.value.real;
+    }
+  };
+  while (!should_exit_) {
+    request_param("MPC_XY_CRUISE", px4_.param_mpc_xy_cruise);
+    request_param("MPC_COL_PREV_D", px4_.param_mpc_col_prev_d);
+    request_param("MPC_LAND_SPEED", px4_.param_mpc_land_speed);
+    request_param("NAV_ACC_RAD", px4_.param_nav_acc_rad);
+
+    if (!std::isfinite(px4_.param_mpc_xy_cruise) || !std::isfinite(px4_.param_mpc_col_prev_d) ||
+        !std::isfinite(px4_.param_mpc_land_speed) || !std::isfinite(px4_.param_nav_acc_rad)) {
+      std::this_thread::sleep_for(std::chrono::seconds(5));
+    } else {
+      std::this_thread::sleep_for(std::chrono::seconds(30));
+    }
+  }
+}
+
+void AvoidanceNode::missionCallback(const mavros_msgs::WaypointList& msg) {
```

Comment:
This gets a little complicated, I think it should be something good to have a unit test or 3 for.

Reply:
done